portfolio/src/main/java/com/google/sps/servlets/FileHandlerServlet.java

```
@@ -0,0 +1,120 @@
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(imageUrls));
+
+    }
+
+    /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. 
+      * (method borrowed from examples)
+      */
+    private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+        List<BlobKey> blobKeys = blobs.get("image");
+
+        // User submitted form without selecting a file, so we can't get a URL. (dev server)
+        if (blobKeys == null || blobKeys.isEmpty()) {
+            return null;
+        }
+
+        // Our form only contains a single file input, so get the first index.
+        BlobKey blobKey = blobKeys.get(0);
+
+        // User submitted form without selecting a file, so we can't get a URL. (live server)
+        // or user did not submit image type 
+        BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+        if (blobInfo.getSize() == 0 || !blobInfo.getContentType().startsWith("image")) {
+            blobstoreService.delete(blobKey);
+            return null;
+        }
+
+        // Use ImagesService to get a URL that points to the uploaded file.
+        ImagesService imagesService = ImagesServiceFactory.getImagesService();
+        ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+        // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+        // path to the image, rather than the path returned by imagesService which contains a host.
+        try {
+            URL url = new URL(imagesService.getServingUrl(options));
+            return url.getPath();
+        } catch (MalformedURLException e) {
+            return imagesService.getServingUrl(options);
```

so your method said it returns a URL, yet this isn't a well formed URL.  I would just throw the exception and let that be the response to the client.

---

Changed to throw exception.

---

Changed to throw the exception.